cartography/intel/gcp/crm.py

```diff
@@ -45,14 +69,158 @@ def load_gcp_organizations(neo4j_session, data, gcp_update_tag):
         )
 
 
+def load_gcp_folders(neo4j_session, data, gcp_update_tag):
+    """
+    Ingest the GCP folders to Neo4j
+    :param neo4j_session: The Neo4j session
+    :param data: List of folders; output from crm.get_gcp_folders()
+    :param gcp_update_tag: The timestamp value to set our new Neo4j nodes with
+    :return: Nothing
+    """
+    for folder in data:
+        # Get the correct parent type.
+        # Parents of folders can only be GCPOrganizations or other folders, see
+        # https://cloud.google.com/resource-manager/docs/cloud-platform-resource-hierarchy
+        if folder['parent'].startswith("organizations"):
+            query = "MATCH (parent:GCPOrganization{id:{ParentId}})"
+        elif folder['parent'].startswith("folders"):
+            query = """
+            MERGE (parent:GCPFolder{id:{ParentId}})
+            ON CREATE SET parent.firstseen = timestamp()
+            """
+        query += """
+        MERGE (folder:GCPFolder{id:{FolderName}})
+        ON CREATE SET folder.firstseen = timestamp()
+        SET folder.foldername = {FolderName},
+        folder.displayname = {DisplayName},
+        folder.lifecyclestate = {LifecycleState},
+        folder.lastupdated = {gcp_update_tag}
+        WITH parent, folder
+        MERGE (parent)-[r:RESOURCE]->(folder)
+        ON CREATE SET r.firstseen = timestamp()
+        SET r.lastupdated = {gcp_update_tag}
+        """
+        neo4j_session.run(
+            query,
+            ParentId=folder['parent'],
+            FolderName=folder['name'],
+            DisplayName=folder.get('displayName', None),
+            LifecycleState=folder.get('lifecycleState', None),
+            gcp_update_tag=gcp_update_tag
+        )
+
+
+def load_gcp_projects(neo4j_session, data, gcp_update_tag):
+    """
+    Ingest the GCP projects to Neo4j
+    :param neo4j_session: The Neo4j session
+    :param data: List of GCP projects; output from crm.get_gcp_projects()
+    :param gcp_update_tag: The timestamp value to set our new Neo4j nodes with
+    :return: Nothing
+    """
+    for project in data:
+        if project['parent']['type'] == "organization":
+            query = "MATCH (parent:GCPOrganization{id:{ParentId}})"
+            parentid = f"organizations/{project['parent']['id']}"
+        elif project['parent']['type'] == "folder":
+            query = """
+            MERGE (parent:GCPFolder{id:{ParentId}})
+            ON CREATE SET parent.firstseen = timestamp()
+            """
+            parentid = f"folders/{project['parent']['id']}"
+        query += """
+        MERGE (project:GCPProject{id:{ProjectId}})
+        ON CREATE SET project.firstseen = timestamp()
+        SET project.projectid = {ProjectId},
+        project.displayname = {DisplayName},
+        project.lifecyclestate = {LifecycleState},
+        project.lastupdated = {gcp_update_tag}
+        WITH parent, project
+        MERGE (parent)-[r:RESOURCE]->(project)
```

Comment:
I don't know if this is the right relationship to represent the folder/organizations contains project concept. That said, I don't have any good alternative to offer. Thoughts?

Reply:
Had it originally as :PARENT since that's what GCP seems to call this link, but it seems that we have :RESOURCE elsewhere to describe parent-child relationships.  I'm open for changes, though I think one benefit of setting all of these parent-child relationships as [:RESOURCE] is that we could query for multiple hops like this:

```cypher
match p=(o:GCPOrganization)-[r:RESOURCE*..]->(g:GCPInstance) return p
```

which would return all instances under an organization regardless of recursion depth.  This only works if we have all these rels as :RESOURCE.